File: go/500.600/539.MinimumTimeDifference.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

func findMinDifference(timePoints []string) int {
	n := len(timePoints)
	if n > 1440 {
		return 0
	}
	dp := []int{}
	for _, v := range timePoints {
		dp = append(dp, toi(&v))
	}
	sort.Ints(dp)

	ans := 1440
	for i := 1; i < n; i++ {
		ans = min(ans, dp[i]-dp[i-1])
	}
	ans = min(ans, 1440-dp[n-1]+dp[0])
	return ans
}

func toi(ts *string) int {
	strs := strings.Split(*ts, ":")
	h, _ := strconv.Atoi(strs[0])
	m, _ := strconv.Atoi(strs[1])
	return h*60 + m
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

func main() {
	fmt.Println(findMinDifference([]string{"23:59", "00:00"}), 1)
	fmt.Println(findMinDifference([]string{"00:00", "23:59", "00:00"}), 0)
	fmt.Println(findMinDifference([]string{"00:00", "04:00", "22:00"}), 120)
}

File: 500.600/539.MinimumTimeDifference.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class MinimumTimeDifference
{
public:
    int findMinDifference(const vector<string> &timePoints)
    {
        int n = timePoints.size();
        if (n > 1440)
        {
            return 0;
        }
        vector<int> dp;
        for (auto &s : timePoints)
        {
            dp.push_back(convertToMin(s));
        }
        sort(dp.begin(), dp.end());
        int ans = 1440;
        for (int i = 1; i < n; ++i)
        {
            int x = dp[i] - dp[i - 1];
            ans = min(ans, x);
        }
        // 特殊情况(最后一个 和 第一个)
        int x = 1440 - dp[n - 1] + dp[0];
        ans = min(ans, x);
        return ans;
    }

private:
    int convertToMin(const string &s)
    {
        int h = atoi(s.c_str());
        int m = atoi(s.substr(s.find(":") + 1).c_str());
        return h * 60 + m;
    }
};

int main()
{
    MinimumTimeDifference cls;
    cout << cls.findMinDifference(vector<string>{"00:00", "04:00", "22:00"}) << " -> 120" << endl;
    cout << cls.findMinDifference(vector<string>{"23:59", "00:00"}) << " -> 1" << endl;
    cout << cls.findMinDifference(vector<string>{"00:00", "23:59", "00:00"}) << " -> 0" << endl;
    return 0;
}
